c/pthread-ext/18_read_write_lock.c

```diff
@@ -37,9 +41,25 @@ void __VERIFIER_atomic_r()
 
 void* thr2(void* arg) { //reader
   __VERIFIER_atomic_r();
-  y = x;
-  assert(y == x);
-  r = r-1;
+  __VERIFIER_atomic_begin();
+  int l = x;
+  __VERIFIER_atomic_end();
+  __VERIFIER_atomic_begin();
+  y = l;
+  __VERIFIER_atomic_end();
+  __VERIFIER_atomic_begin();
+  int ly = y;
+  __VERIFIER_atomic_end();
+  __VERIFIER_atomic_begin();
+  int lx = x;
+  __VERIFIER_atomic_end();
+  assert(ly == lx);
+  __VERIFIER_atomic_begin();
+  int lr = r;
+  __VERIFIER_atomic_end();
+  __VERIFIER_atomic_begin();
+  r = lr-1;
+  __VERIFIER_atomic_end();
```

Comment:
wouldn't this be enough?

```c
  __VERIFIER_atomic_begin();
  y = x;
  __VERIFIER_atomic_end();
  assert(y == x);
  __VERIFIER_atomic_begin();
  int lr = r;
  __VERIFIER_atomic_end();
  __VERIFIER_atomic_begin();
  r = lr-1;
  __VERIFIER_atomic_end();
```

Reply:
No, e.g. the command `y =x` corresponds to two memory accesses (reading the value of `x` and writing that value into `y`). If you put it as a single instruction inside the atomic block, then you don't allow a context switch between the read and the write (which would be allow if no atomic blocks would be used at all, but we need them to remove the data races)